### DISCUSSION TITLE
2.4.0-RC

### DISCUSSION THREAD

**Discussion by mashb1t** [lllyasviel/Fooocus#2955] [General]
## What's Changed
* feat: support download huggingface files from a  mirror site by @chenxinlong in https://github.com/lllyasviel/Fooocus/pull/2637
* chore: update interposer from v3.1 to v4.0 by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2717
* feat: add button to reconnect UI without having to reload the page by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2727
* feat: add optional model VAE select by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2867
* feat: choose random style by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2855
* feat: update anime from animaPencilXL_v100 to animaPencilXL_v310 by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2454
* refactor: rename label for reconnect button by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2893
* feat: add full raw prompt to history log by @docppp in https://github.com/lllyasviel/Fooocus/pull/1920
* fix: use correct border radius css property by @khanvilkarvishvesh in https://github.com/lllyasviel/Fooocus/pull/2845
* fix: do not close meta tag in HTML header by @e52fa787 in https://github.com/lllyasviel/Fooocus/pull/2740
* feat: automatically describe image on uov image upload by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/1938
* add nsfw image censoring via config and checkbox by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/958
* feat: add align your steps scheduler by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2905
* feat: add support for lora inline prompt references by @cantor-set in https://github.com/lllyasviel/Fooocus/pull/2323
* feat: add tcd sampler and discrete distilled tcd scheduler based on sgm_uniform (same as lcm) by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2907
* feat: add performance hyper-sd based on 4step LoRA by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2812

## New Contributors
* @chenxinlong made their first contribution in https://github.com/lllyasviel/Fooocus/pull/2637
* @docppp made their first contribution in https://github.com/lllyasviel/Fooocus/pull/1920
* @khanvilkarvishvesh made their first contribution in https://github.com/lllyasviel/Fooocus/pull/2845
* @e52fa787 made their first contribution in https://github.com/lllyasviel/Fooocus/pull/2740
* @cantor-set made their first contribution in https://github.com/lllyasviel/Fooocus/pull/2323

**Full Changelog**: https://github.com/lllyasviel/Fooocus/compare/2.3.1...2.4.0-rc1

<hr /><em>This discussion was created from the release <a href='https://github.com/lllyasviel/Fooocus/releases/tag/2.4.0-rc1'>2.4.0-rc1</a>.</em>

**Comment by mashb1t** [comment]
Milestone: [2.4.0](https://github.com/lllyasviel/Fooocus/milestone/5)

Info:
- TCD sampler info: eta fixed to 0.3, currently not adjustable in UI.

Open ToDo's: 

- [x] benchmark TCD sampler for hyper-sd LoRA
- [x] extensively compare Hyper-SD CFG LoRAs with non-CFG ones (currently non-cfg ones are better and 4step non-cfg hits the sweet spot IMHO)
- [x] check if [package updates](https://github.com/lllyasviel/Fooocus/milestone/4) should also be included in this release 
- [x] check if NSFW checker can be optimized, currently always loaded in RAM / swap, maybe offloading can be done for low resource scenarios
- [x] Check if inline LoRA references can also be applied per task after wildcard resolution

**Comment by ILikeToasters** [comment]
"add support for lora inline prompt references" Curious if this will allow for usage in wildcards or not? That would be great but I am going to guess maybe not...

**Reply by mashb1t** [reply]
Currently this is not included as LoRA resolving happens before wildcards and LoRAs are handled on a global level, not per task.

**Reply by ILikeToasters** [reply]
That was what I figured but I could hope. lol. Great work by the way. Wish my brain worked as well as it did a few years ago and I could help with the development.

**Reply by mashb1t** [reply]
Let me see what i can do

**Reply by mashb1t** [reply]
@ILikeToasters after multiple hours of code adjustments (see https://github.com/lllyasviel/Fooocus/pull/2966)and debugging this seems not to be possible as the pipeline gets initialized with LoRA and then patched for everything (VAE, ControlNet, etc.), so this sadly is not possible by design.

Nevertheless the time was not fully wasted as some optimisations could be implemented, see https://github.com/lllyasviel/Fooocus/pull/2967

**Reply by ILikeToasters** [reply]
Thanks for making the attempt! I really wasn't expecting you to but really appreciate it that you did try. Glad it was not fully wasted time. I am enjoying the random style feature by the way.

**Comment by Alexdnk** [comment]
tcd sampler with karras/normal/exponential/align_your_steps schedulers got black square at end step. I don't know if it's a bug? The blackness is on the very last step. The other schedulers are ок.

**Reply by mashb1t** [reply]
@Alexdnk this is expected, tcd currently only supports the following schedulers:
- simple
- sgm_uniform
- tcd (sgm_uniform with discrete distinct handling)

also check out the [TCD LoRA](https://huggingface.co/h1t/TCD-SDXL-LoRA/blob/main/pytorch_lora_weights.safetensors) (suggestion: use with tcd sampler and scheduler, weight 0,75 or less)

**Comment by mashb1t** [comment]
Target release date is either 25.05. or 26.05.

**Reply by ILikeToasters** [reply]
I do have one question. The reconnect button, where is that supposed to show?  I wanted to test it out but did not see anything when I closed the CMD window and relaunched Fooocus. Maybe I am missing something obvious... Would not be first time.

**Reply by mashb1t** [reply]
@ILikeToasters the reconnect button shows up instead of the generate button as soon as the websocket connection encounters a timeout, so only if you perform an action in the UI while Fooocus isn't available or when starting to generate and Fooocus (cmd) crashes.

You can test this when you start to generate and close the cmd window.
As soon as an error is encountered the reconmect button shows up and allows to reconnect without having to reload the tab, which would lead to all configuratoons getting lost.
This is also very helpful in developing Fooocus as the browser tab basically only needs to be reloaded after performing gradio changes.

Hope this helps in clarifying the functionality.

**Reply by ILikeToasters** [reply]
Perfectly. And thank you for the detailed explanation.

**Reply by ILikeToasters** [reply]
And it works perfectly. So it seems all things work. Good job!

**Comment by wisdombox44** [comment]
Is it posibble support Stable Cascade and Playground V2.5 models. I tried them on ComfyUI that would be so nice native support on Fooocus.

**Reply by ILikeToasters** [reply]
Some info here:
https://github.com/lllyasviel/Fooocus/discussions/2366
https://github.com/lllyasviel/Fooocus/discussions/2249

**Reply by wisdombox44** [reply]
Thanks, I started playing with StableSwarm it is an interesting WebUI it is natively supported both models. Good thing is you can import ComfyUI workflows, quiete versatile and modular. Whatever WebUI i try Fooocus is still the best WebUI it is simple, easy, an you can go advanced

**Comment by PeterJared** [comment]
Is this an automatic update? Mine says yet 2.3.1
"Already up-to-date
Update succeeded.
[System ARGV] ['Fooocus\\entry_with_update.py', '--preset', 'VLighting']
Python 3.10.9 (tags/v3.10.9:1dd9be6, Dec  6 2022, 20:01:21) [MSC v.1934 64 bit (AMD64)]
Fooocus version: 2.3.1"

**Reply by mashb1t** [reply]
@PeterJared release candidates are not automatically rolled out, but the final version 2.4.0 will. Fooocus checks out the latest main commit on startup, so you'll get it when it's officially published and merged from develop into main, no worries.

**Comment by Madxis** [comment]
With this + the option of changing Clip Skip, Fooocus is going to be perfect!!
Thank you so much for all the hard work!!

**Reply by mashb1t** [reply]
@Madxis CLIP Skip has been implemented in PR https://github.com/lllyasviel/Fooocus/pull/2999

**Comment by ILikeToasters** [comment]
I have a video ready covering all the new stuff once the update goes live. Think I have covered all the details of how it all works.

**Reply by mashb1t** [reply]
@ILikeToasters https://github.com/lllyasviel/Fooocus/issues/2763 will also be included and I try to add clip skip, should be a low hanging fruit (see https://github.com/lllyasviel/Fooocus/discussions/2170#discussioncomment-8432532).

**Reply by mashb1t** [reply]
@ILikeToasters these three things have been added and the release is now feature frozen:
- https://github.com/lllyasviel/Fooocus/pull/2999
- https://github.com/lllyasviel/Fooocus/pull/2998
- https://github.com/lllyasviel/Fooocus/pull/2997

**Reply by ILikeToasters** [reply]
Thanks! Will check those out and add to my list.

**Reply by ILikeToasters** [reply]
I love the new inpaint mask brush feature! All works great!

**Reply by ghost** [reply]
> I love the new inpaint mask brush feature! All works great!

Mine is still showing 2.3.1.

**Reply by ILikeToasters** [reply]
> > I love the new inpaint mask brush feature! All works great!
> 
> Mine is still showing 2.3.1.

Update has not been pushed out yet. According to @mashb1t most likely later today unless that has changed but all seems to be in good working order with the release candidate.

**Reply by mashb1t** [reply]
@Vitali774 it's out now, https://github.com/lllyasviel/Fooocus/releases/tag/v2.4.0

@ILikeToasters one more thing has been added: https://github.com/lllyasviel/Fooocus/pull/2590 & fix in https://github.com/lllyasviel/Fooocus/pull/3001

 Let's continue the discussion in https://github.com/lllyasviel/Fooocus/discussions/3003
